### ISSUE TITLE
Support for dictionaries

### ISSUE THREAD

**Issue by ir2718** [KevinMusgrave/pytorch-metric-learning#734]
Hi,

I'm working on a project that requires me to use multiple models within a single `nn.Module`. However, the preprocessing for the images for one model differs from the others. My initial idea to tackle this was to return a dictionary with preprocessed data from my custom dataset, but unfortunately the [following function](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/utils/common_functions.py#L462) prevents me from moving the data onto a GPU. However, monkey patching this to support dictionaries does not resolve the issue, as other methods also do not support dictionaries.  

So, my question is: is there a way to use the library with the model inputs being dictionaries?

**Comment by KevinMusgrave** [comment]
I think it makes sense that the loss functions, miners, and distances take in tensors. 

Do you mean using dictionaries for the `trainers` and `testers`? Maybe creating a child class with the necessary modifications is the only way to go.

**Comment by ir2718** [comment]
In the meantime, I've found that the issue was on my side. In case anyone else stumbles on a similar use case, you can subclass torch.utils.data.Dataset, return a dictionary for the input, while returning the labels as a tensor and it should work. I'm closing this for now.
